### PR TITLE
suggest parenthesis around ExprWithBlock BinOp ExprWithBlock

### LINKED ISSUE #102171: Missing "parentheses are required to parse this as an expression" hint with multiple unsafe blocks
Given the following [code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ec6fc42ba9cedb3c2a125bea72d12bbd): 

```rust
unsafe fn id(x: u32) -> u32 {
    x
}

fn main() {
    let y = unsafe {
        unsafe { id(3) } + unsafe { id(4) }
    };
    println!("{}", y);
}
```

The current output is:

```
error: expected expression, found `+`
 --> src/main.rs:7:26
  |
7 |         unsafe { id(3) } + unsafe { id(4) }
  |                          ^ expected expression

error[E0308]: mismatched types
 --> src/main.rs:7:18
  |
7 |         unsafe { id(3) } + unsafe { id(4) }
  |                  ^^^^^- help: consider using a semicolon here: `;`
  |                  |
  |                  expected `()`, found `u32`

For more information about this error, try `rustc --explain E0308`.
```

Ideally the output should look like:

```
error: expected expression, found `+`
 --> src/main.rs:7:26
  |
7 |         unsafe { id(3) } + unsafe { id(4) }
  |                          ^ expected expression
  |
help: parentheses are required to parse this as an expression
  |
7 |         (unsafe { id(3) }) + unsafe { id(4) }
  |         +                +
```

Related to #67347

### LINKED ISSUE #105179: match + match compilation error
I tried this code:

```
fn foo(a: u64, b: u64) -> u64 {
    match a {
        0 => 1,
        value => value
    } + match b {
        1 => 0,
        value => value
    }
}

fn main() {
    println!("{}", foo(0, 1));
}
```

I expected to see this happen: the program compiles

Instead, this happened: the program doesn't compile. Note that if you use a return the program compiles and in that case clippy suggests to remove the return.
```
error: expected expression, found `+`
 --> src/main.rs:5:7
  |
5 |     } + match b {
  |       ^ expected expression

error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
 --> src/main.rs:2:5
  |
2 | /     match a {
3 | |         0 => 1,
4 | |         value => value
5 | |     } + match b {
  | |     ^- help: consider using a semicolon here
  | |_____|
  |       expected `()`, found `u64`
```

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

### PR DESCRIPTION
**Pull request by lukas-code**
fix https://github.com/rust-lang/rust/issues/105179
fix https://github.com/rust-lang/rust/issues/102171

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 1]
```diff
@@ -32,11 +32,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
 
     pub(in super::super) fn suggest_semicolon_at_end(&self, span: Span, err: &mut Diagnostic) {
+        // This suggestion is incorrect for
+        // fn foo() -> bool { match () { () => true } || match () { () => true } }
         err.span_suggestion_short(
             span.shrink_to_hi(),
             "consider using a semicolon here",
             ";",
-            Applicability::MachineApplicable,
+            Applicability::MaybeIncorrect,
```
Comment (lukas-code) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:41]: This was already wrong on nightly and isn't caused by this PR.

**Comment by robinmoussu** [comment]
Is this a temporary fix? Shouldn’t `ExprWithBlock BinOp ExprWithBlock` compiles?

I was expecting 

```
match () { 
    () => 1
} + match () {
    () => 2
}
```

 to be valid Rust and equals to `3`.

**Comment by lukas-code** [comment]
> Is this a temporary fix? Shouldn’t `ExprWithBlock BinOp ExprWithBlock` compiles?

This is an unfortunate side effect of the current [Rust grammar](https://doc.rust-lang.org/nightly/reference/statements.html#expression-statements). When a statement starts with an *ExpressionWithBlock* (such as a `match` expression), the statement ends immediately after the closing `}`.

Changing this will probably require an [RFC](https://rustc-dev-guide.rust-lang.org/walkthrough.html#pre-rfc-and-rfc) and a nontrivial amount of work, since this would introduce parsing ambiguities based on types, [for example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3e100ddcfd044ec842cbf4069e54778d):
```rust
fn foo() -> impl FnOnce() -> bool {
    match () { () => () }
    || match () { () => true } // closure
}

fn bar() -> bool {
    // (match () { () => true }) || match () { () => true } // logical or
    match () { () => true } || match () { () => true } // logical or
}

fn baz() -> &'static &'static bool {
    match () { () => () }
    &&match () { () => true } // reference to reference
}

fn quux() -> bool {
    // (match () { () => true }) && match () { () => true } // logical and
    match () { () => true } && match () { () => true } // logical and
}
```

Furthermore, this would break resolution of unary vs binary `-` and should probably be gated by an edition. [Example:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f602ed16321d054c9d4e362bd6fcb56c)
```rust
use std::ops;

struct Bar;
impl ops::Neg for Bar {
    type Output = ();

    fn neg(self) {
        println!("neg");
    }
}
impl ops::Sub<Bar> for () {
    type Output = ();

    fn sub(self, _: Bar) {
        println!("sub");
    }
}

fn main() {
    {
        match () {
            () => (),
        } - match () {
            () => Bar,
        }
    }

    {
        (match () {
            () => (),
        }) - match () {
            () => Bar,
        }
    }
}
```

**Comment by petrochenkov** [comment]
@borsr +

**Comment by robinmoussu** [comment]
> This is an unfortunate side effect of the current [Rust grammar](https://doc.rust-lang.org/nightly/reference/statements.html#expression-statements). When a statement starts with an _ExpressionWithBlock_ (such as a `match` expression), the statement ends immediately after the closing `}`.

Basically, this would require to have significant newline to distinguish them. Or to require a `;` at the end of each statement, include `match`, `if` and similar statement that have a pair of braces. Yeah, it doesn’t sound great.

**Comment by robinmoussu** [comment]
> This is an unfortunate side effect of the current [Rust grammar](https://doc.rust-lang.org/nightly/reference/statements.html#expression-statements). When a statement starts with an _ExpressionWithBlock_ (such as a `match` expression), the statement ends immediately after the closing `}`.

Basically, this would require to have significant newline to distinguish them. Or to require a `;` at the end of each statement, include `match`, `if` and similar statement that have a pair of braces. Yeah, it doesn’t sound great.

**Pull request merged**